src/logging/LogQueue.h

```diff
@@ -0,0 +1,95 @@
+/*
+	SlimeVR Code is placed under the MIT license
+	Copyright (c) 2024 Jabberrock & SlimeVR contributors
+
+	Permission is hereby granted, free of charge, to any person obtaining a copy
+	of this software and associated documentation files (the "Software"), to deal
+	in the Software without restriction, including without limitation the rights
+	to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+	copies of the Software, and to permit persons to whom the Software is
+	furnished to do so, subject to the following conditions:
+
+	The above copyright notice and this permission notice shall be included in
+	all copies or substantial portions of the Software.
+
+	THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+	IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+	FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+	AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+	LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+	OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+	THE SOFTWARE.
+*/
+
+#ifndef LOGGING_LOGQUEUE_H
+#define LOGGING_LOGQUEUE_H
+
+#include <array>
+#include <cstdint>
+#include <cstring>
+#include <numeric>
+
+namespace SlimeVR::Logging {
+
+class LogQueue {
+public:
+	// Whether there are any messages in the queue.
+	bool empty() const;
+
+	// First message in the queue.
+	const char* front() const;
+
+	// Adds a message to the end of the queue. Messages that are too long will be truncated.
+	void push(const char* message);
+
+	// Removes a message from the front of the queue.
+	void pop();
+
+	// Global instance of the log queue
+	static LogQueue& instance();
+
+private:
+	// Sets the message at a particular offset
+	void setMessageAt(int offset, const char* message);
+
+	static constexpr size_t MaxMessages = 100;
+	static constexpr size_t MaxMessageLength = std::numeric_limits<uint8_t>::max();
+	static constexpr char OverflowMessage[] = "[OVERFLOW]";
+
+	template <typename T, T Modulus>
+	class Modulo {
```

Comment:
It's a private class at the moment. I didn't want to move it out because it doesn't support all the necessary functions, nor does it take care of general edge cases.

Reply:
I think I'll just get rid of the helper class and move the modulo logic into the `LogQueue` implementation.

Reply:
Addressed this by dropping the Module helper class.